backend/danswer/direct_qa/local_llm.py

```diff
@@ -0,0 +1,89 @@
+import re
+from collections.abc import Generator
+from typing import Any, Optional, Union
+
+from transformers import pipeline  # type:ignore
+from transformers import QuestionAnsweringPipeline  # type:ignore
+
+from danswer.chunking.models import InferenceChunk
+from danswer.configs.constants import BLURB
+from danswer.configs.constants import DOCUMENT_ID
+from danswer.configs.constants import SEMANTIC_IDENTIFIER
+from danswer.configs.constants import SOURCE_LINK
+from danswer.configs.constants import SOURCE_TYPE
+from danswer.configs.model_configs import LOCAL_LLM_MODEL
+from danswer.direct_qa.interfaces import QAModel
+from danswer.utils.logger import setup_logger
+from danswer.utils.timing import log_function_time
+
+# Use a pipeline as a high-level helper
+
+logger = setup_logger()
+
+_QA_MODEL = None
+
+
+def get_default_qa_model() -> QuestionAnsweringPipeline:
+    global _QA_MODEL
+    if _QA_MODEL is None:
+        _QA_MODEL = pipeline("question-answering", model=LOCAL_LLM_MODEL)
+
+    return _QA_MODEL
+
+
+def strip_context_document(document: str) -> str:
+    return re.sub(r"[\n\r\s]+", " ", document)
+
+
+class LocalModel(QAModel):
+    @log_function_time()
+    def answer_question(
+        self, query: str, context_docs: list[InferenceChunk]
+    ) -> tuple[str | None, dict[str, dict[str, str | int | None]] | None]:
+        quotes_dict: dict[str, dict[str, Union[str, int, None]]] = {}
+        top_contents = [
+            strip_context_document(ranked_chunk.content)
```

Comment:
This works because I do not do any promts. This model can be much simpler because it gets the context in chunks and the question independently. It will only find the best quote in the context which fits the question. Given that your question is answered on one of the chunks found be semantic search this works well.

I did some improvements yesterday evening to always include the complete sentence. Otherwise it will often only return a few works. Things like: "How to I send mails?" -> "AWS Simple Email Service". While technically correct it would be nice to have the full sentence such as "To send emails in your application can use AWS Simple Email Service by doing X and Y."